src/hooks/useCopyToClipboard.ts

```diff
@@ -0,0 +1,333 @@
+      "dyad-rename",
+      "dyad-delete",
+      "dyad-add-dependency",
+      "dyad-execute-sql",
+      "dyad-add-integration",
+      "dyad-output",
+      "dyad-problem-report",
+      "dyad-chat-summary",
+      "dyad-edit",
+      "dyad-codebase-context",
+      "think",
+      "dyad-command",
+    ];
+
+    const tagPattern = new RegExp(
+      `<(${customTagNames.join("|")})\\s*([^>]*)>(.*?)<\\/\\1>`,
+      "gs",
+    );
+
+    const contentPieces: any[] = [];
+    let lastIndex = 0;
+    let match;
+
+    while ((match = tagPattern.exec(processedContent)) !== null) {
+      const [fullMatch, tag, attributesStr, tagContent] = match;
+      const startIndex = match.index;
+
+      // Add markdown content before this tag
+      if (startIndex > lastIndex) {
+        contentPieces.push({
+          type: "markdown",
+          content: processedContent.substring(lastIndex, startIndex),
+        });
+      }
+
+      // Parse attributes
+      const attributes: Record<string, string> = {};
+      const attrPattern = /(\w+)="([^"]*)"/g;
+      let attrMatch;
+      while ((attrMatch = attrPattern.exec(attributesStr)) !== null) {
+        attributes[attrMatch[1]] = attrMatch[2];
+      }
+
+      // Add the tag info
+      contentPieces.push({
+        type: "custom-tag",
+        tagInfo: {
+          tag,
+          attributes,
+          content: tagContent,
+          fullMatch,
+        },
+      });
+
+      lastIndex = startIndex + fullMatch.length;
+    }
+
+    // Add remaining markdown content
+    if (lastIndex < processedContent.length) {
+      contentPieces.push({
+        type: "markdown",
+        content: processedContent.substring(lastIndex),
+      });
+    }
+
+    return contentPieces;
+  };
+
+  // Simplified version of preprocessUnclosedTags
+  const preprocessUnclosedTags = (content: string) => {
+    const customTagNames = [
+      "dyad-write",
+      "dyad-rename",
+      "dyad-delete",
+      "dyad-add-dependency",
+      "dyad-execute-sql",
+      "dyad-add-integration",
+      "dyad-output",
+      "dyad-problem-report",
+      "dyad-chat-summary",
+      "dyad-edit",
+      "dyad-codebase-context",
+      "think",
+      "dyad-command",
+    ];
+
+    let processedContent = content;
+
+    for (const tagName of customTagNames) {
+      const openTagPattern = new RegExp(`<${tagName}(?:\\s[^>]*)?>`, "g");
+      const closeTagPattern = new RegExp(`</${tagName}>`, "g");
+
+      const openCount = (processedContent.match(openTagPattern) || []).length;
+      const closeCount = (processedContent.match(closeTagPattern) || []).length;
+
+      const missingCloseTags = openCount - closeCount;
+      if (missingCloseTags > 0) {
+        processedContent += Array(missingCloseTags)
+          .fill(`</${tagName}>`)
+          .join("");
+      }
+    }
+
+    return { processedContent };
+  };
+
+  // Detect programming language from file extension
+  const detectLanguage = (filename: string): string => {
```

Comment:
let's extract this into a shared util file and reuse with https://github.com/dyad-sh/dyad/blob/a547aa3ac1253fa649f7e546cdb996026f91c3fb/src/components/preview_panel/FileEditor.tsx#L196